File: authorized-charges-never-captured/python/capture_authorized.py
"""Capture WooCommerce orders whose Stripe authorization was never captured.

Manual-capture orders sit on hold with an authorized-but-uncaptured PaymentIntent.
Stripe holds an authorization for about 7 days, then releases it and the money is
gone. This finds authorized PaymentIntents whose order is still on hold, checks the
amount, and captures them before the hold expires.

Run on a schedule. Safe to run again and again.
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("capture_authorized")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "168"))  # 7 days
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CAPTURABLE_STATUSES = {"on-hold", "pending"}


def recent_uncaptured(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    for intent in stripe.PaymentIntent.list(limit=100, created={"gte": since}).auto_paging_iter():
        if intent.status == "requires_capture" and intent.metadata.get("order_id"):
            yield intent


def get_order(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def order_amount_minor(order):
    return round(float(order["total"]) * 100)


def decide(order, intent):
    # An authorized PaymentIntent reports the held amount in `amount`, not
    # `amount_received`, because nothing has been captured yet.
    if intent.get("status") != "requires_capture":
        return ("skip", "not awaiting capture")
    if order is None:
        return ("orphan", "order not found")
    if order["status"] not in CAPTURABLE_STATUSES:
        return ("skip", "order not awaiting capture")
    if abs(order_amount_minor(order) - intent["amount"]) > 1:
        return ("mismatch", "amount does not match")
    return ("capture", "authorized in Stripe, still awaiting capture")


def capture(order_id, intent):
    charge = stripe.PaymentIntent.capture(intent["id"])
    charge_id = charge.get("latest_charge") or intent["id"]
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}",
        json={"status": "processing", "transaction_id": charge_id},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": f"Captured Stripe PaymentIntent {intent['id']} before the "
                      f"authorization expired. Marked processing by the capture job."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    captured = 0
    for intent in recent_uncaptured(LOOKBACK_HOURS):
        order_id = intent.metadata["order_id"]
        order = get_order(order_id)
        action, reason = decide(order, intent)
        if action == "orphan":
            log.warning("Intent %s points to order %s which is missing", intent.id, order_id)
            continue
        if action in ("skip", "mismatch"):
            if action == "mismatch":
                log.warning("Order %s amount mismatch: %s", order_id, reason)
            continue
        log.info("Order %s: %s. %s", order_id, reason, "would capture" if DRY_RUN else "capturing")
        if not DRY_RUN:
            capture(order_id, intent)
        captured += 1
    log.info("Done. %d order(s) %s.", captured, "to capture" if DRY_RUN else "captured")


if __name__ == "__main__":
    run()

File: authorized-charges-never-captured/node/capture-authorized.js
/**
 * Capture WooCommerce orders whose Stripe authorization was never captured.
 *
 * Manual-capture orders sit on hold with an authorized-but-uncaptured PaymentIntent.
 * Stripe holds an authorization for about 7 days, then releases it. This finds
 * authorized PaymentIntents whose order is still on hold, checks the amount, and
 * captures them before the hold expires. Run on a schedule. Safe to run again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/authorized-charges-never-captured/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 168);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const CAPTURABLE_STATUSES = new Set(["on-hold", "pending"]);

export function orderAmountMinor(order) {
  return Math.round(parseFloat(order.total) * 100);
}

export function decide(order, intent) {
  // An authorized PaymentIntent reports the held amount in `amount`, not
  // `amount_received`, because nothing has been captured yet.
  if (intent.status !== "requires_capture") return ["skip", "not awaiting capture"];
  if (!order) return ["orphan", "order not found"];
  if (!CAPTURABLE_STATUSES.has(order.status)) return ["skip", "order not awaiting capture"];
  if (Math.abs(orderAmountMinor(order) - intent.amount) > 1) {
    return ["mismatch", "amount does not match"];
  }
  return ["capture", "authorized in Stripe, still awaiting capture"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* recentUncaptured(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  for await (const intent of stripe.paymentIntents.list({ limit: 100, created: { gte: since } })) {
    if (intent.status === "requires_capture" && intent.metadata.order_id) yield intent;
  }
}

async function capture(orderId, intent) {
  const charge = await stripe.paymentIntents.capture(intent.id);
  const chargeId = charge.latest_charge || intent.id;
  await woo(`/orders/${orderId}`, {
    method: "PUT",
    body: JSON.stringify({ status: "processing", transaction_id: chargeId }),
  });
  await woo(`/orders/${orderId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Captured Stripe PaymentIntent ${intent.id} before the authorization ` +
            `expired. Marked processing by the capture job.`,
    }),
  });
}

export async function run() {
  let captured = 0;
  for await (const intent of recentUncaptured(LOOKBACK_HOURS)) {
    const orderId = intent.metadata.order_id;
    const order = await woo(`/orders/${orderId}`);
    const [action, reason] = decide(order, intent);
    if (action === "orphan") { console.warn(`Intent ${intent.id} points to missing order ${orderId}`); continue; }
    if (action === "skip" || action === "mismatch") {
      if (action === "mismatch") console.warn(`Order ${orderId} amount mismatch: ${reason}`);
      continue;
    }
    console.log(`Order ${orderId}: ${reason}. ${DRY_RUN ? "would capture" : "capturing"}`);
    if (!DRY_RUN) await capture(orderId, intent);
    captured++;
  }
  console.log(`Done. ${captured} order(s) ${DRY_RUN ? "to capture" : "captured"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
